src/number_theory/fermat_psp.lean

```diff
@@ -0,0 +1,548 @@
+  { have h₁ : 2*(a + 1)/(a + 1) ≤ (a ^ b + 1)/(a + 1) := nat.div_le_div_right h,
+    have h₂ : 0 < a + 1 := nat.succ_pos a,
+    rwa nat.mul_div_cancel _ h₂ at h₁ },
+  rw [mul_add, mul_one],
+
+  -- Because `2 ≤ a` and `2 < b`, `3 ≤ a^(b - 1)`
+  have h₁ : 3 ≤ a^(b - 1),
+  { have : 2 ≤ b - 1 := nat.le_pred_of_lt hb,
+    calc a^(b - 1) ≥ a^2 : (pow_le_pow_iff ha).mpr this
+               ... ≥ 2^2 : pow_le_pow_of_le_left' ha₁ 2
+               ... ≥ 3   : by norm_num },
+
+  -- Since we know that `3 ≤ a ^ (b - 1)`, if we want to show `2 * a + 1 ≤ a ^ b` then it suffices
+  -- to show that `2 * a + 1 ≤ 3 * a` because then `2 * a + 1 ≤ a * 3 ≤ a * a^(b - 1) = a ^ b`
+  rw [←nat.sub_add_cancel hb₁, pow_succ],
+  suffices h : 2 * a + 1 ≤ a * a^(b - 1),
+  { exact nat.succ_le_succ h },
+  suffices h : 2 * a + 1 ≤ a * 3,
+  { exact le_mul_of_le_mul_left h h₁ },
+  rw mul_comm a 3,
+
+  -- Because `1 ≤ a`, `a + 1 ≤ 3 * a`
+  have : 3 * a = 2 * a + a := add_one_mul 2 a,
+  rw this,
+  have h : 1 ≤ a := le_of_lt ha,
+  exact add_le_add_left h (2 * a)
+end
+
+private lemma AB_id_helper (b p : ℕ) (hb : 2 ≤ b) (hp : odd p)
+  : ((b ^ p - 1) / (b - 1)) * ((b ^ p + 1) / (b + 1)) = ((b ^ (2*p)) - 1) / (b^2 - 1) :=
+have q₁ : (b - 1) ∣ (b ^ p - 1) := sub_one_dvd_pow_sub_one _ _ (show 1 ≤ b, by linarith),
+have q₂ : (b + 1) ∣ (b ^ p + 1) := begin
+  have : odd (p / 1) := eq.symm (nat.div_one p) ▸ hp,
+  have h := odd_pow_lem ↑b p 1 (one_dvd p) this,
+  rw pow_one at h,
+  exact_mod_cast h,
+end,
+have q₃ : 1 ≤ (b^p) := nat.one_le_pow p b (show 0 < b, by linarith),
+calc (b ^ p - 1) / (b - 1) * ((b ^ p + 1) / (b + 1)) =
+    (b ^ p - 1) * (b ^ p + 1) / ((b - 1) * (b + 1)) : nat.div_mul_div_comm q₁ q₂
+  ... = ((b ^ p + 1) * (b ^ p - 1)) / ((b - 1) * (b + 1)) : by rw mul_comm
+  ... = ((b ^ p)^2 - 1^2) / ((b - 1) * (b + 1))           : by rw nat.sq_sub_sq
+  ... = ((b ^ (p*2)) - 1^2) / ((b - 1) * (b + 1))         : by rw pow_mul
+  ... = ((b ^ (2*p)) - 1^2) / ((b - 1) * (b + 1))         : by rw mul_comm
+  ... = ((b ^ (2*p)) - 1^2) / ((b + 1) * (b - 1))         : by rw mul_comm (b + 1)
+  ... = ((b ^ (2*p)) - 1^2) / (b^2 - 1^2)                 : by rw nat.sq_sub_sq
+  ... = ((b ^ (2*p)) - 1) / (b^2 - 1)                     : by rw one_pow
+
+end helper_lemmas
+
+/--
+Given a prime `p` which does not divide `b*(b^2 - 1)`, we can produce a number `n` which is larger
+than `p` and pseudoprime to base `b`. We do this by defining
+`n = ((b^p - 1)/(b - 1)) * ((b^p + 1)/(b + 1))`
+
+The primary purpose of this definition is to help prove `exists_infinite_pseudoprimes`. For a proof
+that `n` is actually pseudoprime to base `b`, see `psp_from_prime_psp`, and for a proof that `n` is
+greater than `p`, see `psp_from_prime_gt_p`.
+-/
+private def psp_from_prime (b : ℕ) (b_ge_two : 2 ≤ b) (p : ℕ) (p_prime : nat.prime p)
+  (p_gt_two : 2 < p) (not_dvd : ¬p ∣ b*(b^2 - 1)) : ℕ :=
+have A : ℕ := (b^p - 1)/(b - 1),
+have B : ℕ := (b^p + 1)/(b + 1),
+A * B
```

Comment:
If you `#print psp_from_prime` you'll see that the `have`s are somehow embedded into the definition, which is not ideal. You want to make definitions as uncluttered as possible as this makes them easier to use.
```suggestion
private def psp_from_prime (b : ℕ) (b_ge_two : 2 ≤ b) (p : ℕ) (p_prime : nat.prime p)
  (p_gt_two : 2 < p) (not_dvd : ¬p ∣ b*(b^2 - 1)) : ℕ := (b^p - 1)/(b - 1)*((b^p + 1)/(b + 1))
```
Also, I would *strongly* recommend removing `b_ge_two`, `p_prime`, `p_gt_two` and `not_dvd` from this declaration. This is a *definition*, not a theorem. Removing the hypotheses will make the definition much easier to use and will probably shorten your proofs. Hypotheses go in theorems not definitions in Lean -- this is why 1 / 0 = 0 (no hypothesis that denominator is non-zero) but there are no theorems about a / 0 (hypotheses exist in the theorems).

Reply:
I removed the `have`s, and also removed the proofs from the hypothesis and added them as notes to the docstring.